src/node_api_embedding.cc

```diff
@@ -0,0 +1,765 @@
+  v8::MaybeLocal<v8::Value> ret =
+      embedded_env->snapshot()
+          ? node::LoadEnvironment(node_env, node::StartExecutionCallback{})
+          : node::LoadEnvironment(node_env,
+                                  std::string_view(main_script),
+                                  embedded_env->options().preload_cb_);
+
+  embedded_env.release();
+
+  if (ret.IsEmpty()) return napi_pending_exception;
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL node_api_delete_env(napi_env env, int* exit_code) {
+  CHECK_ENV(env);
+  v8impl::EmbeddedEnvironment* embedded_env =
+      v8impl::EmbeddedEnvironment::FromNapiEnv(env);
+
+  if (embedded_env->IsScopeOpened()) return napi_generic_failure;
+
+  {
+    v8impl::IsolateLocker isolate_locker(embedded_env->env_setup());
+
+    int ret = node::SpinEventLoop(embedded_env->node_env()).FromMaybe(1);
+    if (exit_code != nullptr) *exit_code = ret;
+  }
+
+  std::unique_ptr<node::CommonEnvironmentSetup> env_setup =
+      embedded_env->ResetEnvSetup();
+
+  if (embedded_env->create_snapshot()) {
+    node::EmbedderSnapshotData::Pointer snapshot = env_setup->CreateSnapshot();
+    assert(snapshot);
+    embedded_env->create_snapshot()(snapshot.get());
+  }
+
+  node::Stop(embedded_env->node_env());
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL node_api_open_env_scope(napi_env env) {
+  CHECK_ENV(env);
+  v8impl::EmbeddedEnvironment* embedded_env =
+      v8impl::EmbeddedEnvironment::FromNapiEnv(env);
+
+  return embedded_env->OpenScope();
+}
+
+napi_status NAPI_CDECL node_api_close_env_scope(napi_env env) {
+  CHECK_ENV(env);
+  v8impl::EmbeddedEnvironment* embedded_env =
+      v8impl::EmbeddedEnvironment::FromNapiEnv(env);
+
+  return embedded_env->CloseScope();
+}
+
+napi_status NAPI_CDECL node_api_run_env(napi_env env) {
+  CHECK_ENV(env);
+  v8impl::EmbeddedEnvironment* embedded_env =
+      v8impl::EmbeddedEnvironment::FromNapiEnv(env);
+
+  if (node::SpinEventLoopWithoutCleanup(embedded_env->node_env()).IsNothing()) {
+    return napi_closing;
+  }
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL node_api_run_env_while(napi_env env,
+                                              node_api_run_predicate predicate,
+                                              void* predicate_data,
+                                              bool* has_more_work) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, predicate);
+  v8impl::EmbeddedEnvironment* embedded_env =
+      v8impl::EmbeddedEnvironment::FromNapiEnv(env);
+
+  if (predicate(predicate_data)) {
+    if (node::SpinEventLoopWithoutCleanup(
+            embedded_env->node_env(),
+            [predicate, predicate_data]() { return predicate(predicate_data); })
+            .IsNothing()) {
+      return napi_closing;
+    }
+  }
+
+  if (has_more_work != nullptr) {
+    *has_more_work = uv_loop_alive(embedded_env->node_env()->event_loop());
+  }
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL node_api_await_promise(napi_env env,
```

Comment:
At first I thought this might be an extension to the promise support we already have - https://nodejs.org/api/n-api.html#promises

This is a good example were I think we needed the embed or something else in the name as otherwise people might get confused and think it could be called from an addon.

Reply:
or maybe the prefix should be `node_embedding_api_XXX`

Reply:
What if we shorten it to the `node_embedding_`?

Reply:
I'm good with `node_embedding_`

Reply:
All API is changed to use the `napi_embedding_` prefix.